src/core_functions/scalar/list/list_reduce.cpp

```diff
@@ -0,0 +1,241 @@
+
+		idx_t old_count = 0;
+		idx_t new_count = 0;
+
+		// Initialize the left vector from list entries and the active rows
+		auto data = active_rows.GetData();
+
+		for (idx_t entry_idx = 0; old_count < info.row_count; entry_idx++) {
+			if (data[entry_idx] == 0) {
+				old_count += 64;
+				continue;
+			}
+
+			for (idx_t j = 0; j < 64 && entry_idx + j < info.row_count; j++) {
+				auto list_column_format_index = info.list_column_format.sel->get_index(old_count);
+				if (info.list_column_format.validity.RowIsValid(list_column_format_index)) {
+					if (info.list_entries[list_column_format_index].length == 0) {
+						throw ParameterNotAllowedException("Cannot perform list_reduce on an empty input list");
+					}
+					left_vector.set_index(new_count, info.list_entries[list_column_format_index].offset);
+					new_count++;
+				} else {
+					// Remove the invalid rows
+					info.result_validity->SetInvalid(old_count);
+					active_rows.SetInvalid(old_count);
+				}
+				old_count++;
+			}
+		}
+
+		left_slice.Slice(left_vector, new_count);
+
+		if (info.has_index) {
+			input_types.push_back(LogicalType::BIGINT);
+		}
+		input_types.push_back(left_slice.GetType());
+		input_types.push_back(left_slice.GetType());
+		for (auto &entry : info.column_infos) {
+			input_types.push_back(entry.vector.get().GetType());
+		}
+
+		expr_executor = make_uniq<ExpressionExecutor>(context, *info.lambda_expr);
+	};
+	ValidityMask active_rows;
+	Vector left_slice;
+	unique_ptr<ExpressionExecutor> expr_executor;
+	vector<LogicalType> input_types;
+
+	SelectionVector right_sel;
+	SelectionVector left_sel;
+	SelectionVector active_rows_sel;
+};
+
+static bool ExecuteReduce(idx_t loops, ReduceExecuteInfo &execute_info, LambdaFunctions::LambdaInfo &info,
+                          DataChunk &result_chunk) {
+	idx_t old_count = 0;
+	idx_t new_count = 0;
+	idx_t valid_count = 0;
+
+	// create selection vectors for the left and right slice
+	auto data = execute_info.active_rows.GetData();
+
+	for (idx_t entry_idx = 0; old_count < info.row_count; entry_idx++) {
+		if (data[entry_idx] == 0) {
+			old_count += 64;
+			continue;
+		}
+
+		for (idx_t j = 0; j < 64 && entry_idx + j < info.row_count; j++) {
+			if (!execute_info.active_rows.RowIsValid(old_count)) {
+				old_count++;
+				continue;
+			}
+			auto list_column_format_index = info.list_column_format.sel->get_index(old_count);
+			if (info.list_entries[list_column_format_index].length > loops + 1) {
+				execute_info.right_sel.set_index(new_count,
+				                                 info.list_entries[list_column_format_index].offset + loops + 1);
+				execute_info.left_sel.set_index(new_count, valid_count);
+				execute_info.active_rows_sel.set_index(new_count, old_count);
+
+				new_count++;
+			} else {
+				execute_info.active_rows.SetInvalid(old_count);
+				info.result.SetValue(old_count, execute_info.left_slice.GetValue(valid_count));
+			}
+			old_count++;
+			valid_count++;
```

Comment:
Is there a difference between `valid_count` and `old_count`?

Reply:
`old_count` keeps track of all the rows whereas `valid_count` only keeps track of the valid rows. `old_count` gets incremented regardless of whether the row is valid or not.

Reply:
Ah! I get it now... Maybe my confusion indicates that we should rename the variables? What do you think? 

Maybe `old_count` -> `original_row_idx` or `result_row_idx`? And `new_count` -> `reduced_row_idx`? Not sure about valid count. But feel free to also keep the old names, and maybe add a comment.

Reply:
Sure! Good point :)